### ISSUE TITLE
Support for setting access policies on storage containers

### ISSUE THREAD

**Issue by chpspiir** [hashicorp/terraform-provider-azurerm#3722]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Managing retention policies (immutability and legal holds) on storage containers is not yet supported by azurerm_storage_container. It seems that only ACL policies are supported.

It would be great to manage these policies at the infrastructure-level rather than at code level.

I looked at the API docs for the blob service, and I am having trouble figuring out what operation to use:
https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-rest-api. 

Is this because it isn't exposed in the REST API? Or is it a naming confusion that I am not aware of?

If someone could point me in the right direction, I would gladly take a stab at implementing it. 

### New or Affected Resource(s)

* azurerm_storage_container

### Potential Terraform Configuration

```hcl
resource "azurerm_storage_container" "test" {
  name                  = "vhds"
  resource_group_name   = "${azurerm_resource_group.test.name}"
  storage_account_name  = "${azurerm_storage_account.test.name}"
  container_access_type = "private"

  retention_policy {
    type = "legal-hold/immutable-retention" # It might be up for debate whether legal hold and immutable retention should get their own blocks. Mostly for UX.
    retention_time = "7d"
  }
}
```

### References

https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-rest-api. 

* #0000

**Comment by vbrinza** [comment]
(y)

**Comment by tombuildsstuff** [comment]
hey @chpspiir 

Thanks for opening this issue.

This functionality lives within the Data Plane API's for Azure Storage, rather than within the Resource Manager API's - which as such needs a new version of the Azure Storage SDK [than the version we're currently using](http://github.com/Azure/azure-sdk-for-go/tree/master/storage). Unfortunately we're unable to use [the replacement SDK](https://github.com/Azure/azure-storage-blob-go) - since we've been blocked on this for a while (and this is now blocking a large number of issues) we've ended up writing our own SDK, which I'm just finishing up work on at the moment.

Once that's finished (and integrated) - it should be possible to implement this, since I can confirm the new SDK includes support for both of these fields.

Thanks!

**Comment by chpspiir** [comment]
Sounds great! I haven't worked with terraform providers before, but I am up for giving it a stab when the SDK changes are ready. Let me know if I can help :)

**Comment by landro** [comment]
hey @tombuildsstuff , do you have a best-guess ETA for the integration of the new storage sdk you're working on?

**Comment by mamoit** [comment]
> Once that's finished (and integrated) - it should be possible to implement this, since I can confirm the new SDK includes support for both of these fields.

@tombuildsstuff How is the progress on the new SDK? Does it still block this issue?

**Comment by tombuildsstuff** [comment]
@landro / @mamoit the [new Storage SDK's](https://github.com/tombuildsstuff/giovanni) been merged and integrated - however it looks like this needs to be added to the Container's API [since this has been added since the SDK was written](https://docs.microsoft.com/en-us/rest/api/storagerp/blobcontainers/setlegalhold) for that to be added

**Comment by jblafage** [comment]
@tombuildsstuff : do you think if the following kind of REST api around Azure Storage could take part of this [Storage SDK](https://github.com/tombuildsstuff/giovanni)??

The idea is to be able to create a stored access policy for a given container and then generate a sas key based on this access policy. If it could be managed over Terraform it could facilitate implementations. The main advantage using stored access policies is that we can revoke all generated SAS keys based on a given stored access policy.

**Comment by jacksinclairaudley** [comment]
Our terraform configuration would massively benefit from this from a security perspective. At the moment we're struggling how to find a workaround to this so our SAS keys can be created under access policies.

**Comment by ariciputi** [comment]
Hi @tombuildsstuff,
any news on this? It would be nice to have it added either to the Blob Container or to the Storage Account. Not having this it's a showstopper for the Terraform's adoption in the regulated environment where I work.

**Comment by alexwlchan** [comment]
It’s no substitute for being able to define legal holds and the like in Terraform, but now that Terraform 0.13 highlights changes to outputs, we've added an output for "is a legal hold defined on our Blob Storage container":

```console
$ terraform plan -out=terraform.plan
…

Changes to Outputs:
  ~ azure_has_legal_hold = true -> false

------------------------------------------------------------------------

This plan was saved to: terraform.plan

To perform exactly these actions, run the following command to apply:
    terraform apply "terraform.plan"

Releasing state lock. This may take a few moments...
```

It means we're more likely to notice if the immutability policy or legal hold goes walkabout.

**Comment by Furze** [comment]
Hey, is there any updates on if this feature will be added ?

**Comment by tombuildsstuff** [comment]
Blocked on #10765

**Comment by petrkoutnycz** [comment]
So is this thread about container access or retention policy? I am quite confused.

**Comment by nicoarb** [comment]
Are there any news? We are missing the ability to create container access policies and generate sas tokens heavily.

**Comment by shadow100000** [comment]
Ouch

**Comment by alexismosquera** [comment]
Hi!.
Any news about this issue?.

**Comment by psddp** [comment]
The @kensykora solution is great but the command will run only once. 
For some cases (like this policy), you want to make sure the command runs on updates also, to make sure it is recreated if missing (manually deleted). 

A workaround can be to add a new trigger that contains a timestamp: 
```terraform
resource "null_resource" "immutability_policy" {
  triggers = {
    always_run           = "${timestamp()}"
    resource_group_name  = azurerm_storage_account.ops_security.resource_group_name
    storage_account_name = azurerm_storage_account.ops_security.name
    container_name       = azurerm_storage_container.audit_logs.name
    subscription_id      = data.azurerm_subscription.ops.subscription_id
  }

  provisioner "local-exec" {
    command = <<BASH
      az storage container immutability-policy create \
        --resource-group ${self.triggers.resource_group_name} \
        --account-name ${self.triggers.storage_account_name} \
        --container-name ${self.triggers.container_name} \
        --subscription ${self.triggers.subscription_id} \
        --period 365 # days
    BASH
  }
}
```

**Comment by kensykora** [comment]
@psddp a better workaround these days might be to use the [azapi provider](https://registry.terraform.io/providers/azure/azapi/latest/docs) instead which (i don't think) was available when i wrote that workaround. These days for addressing azurerm provider gaps, `azapi` it's my go-to.

**Comment by clowa** [comment]
Here is an example of the solution mentioned by @kensykora. Thanks a lot! I wasn't aware of the [azapi provider](https://registry.terraform.io/providers/azure/azapi/latest/docs) before.


```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = ">=3.4.0"
    }
    azapi = {
      source  = "azure/azapi"
      version = ">=0.1.1"
    }
    random = {
      source  = "random"
      version = ">=3.1.3"
    }
  }
}

provider "azurerm" {
  features {}
  skip_provider_registration = true
}

provider "azapi" {
  skip_provider_registration = true
}

variable "app_name" {
  type = string
}

variable "tags" {
  type = map(string)
}

locals {
  tags = merge({
    "managed-by" = "terraform"
  }, var.tags)
}

resource "random_id" "this" {
  byte_length = 4
}

resource "azurerm_resource_group" "this" {
  name     = upper("RG-${var.app_name}")
  location = "Germany West Central"
}

resource "azurerm_storage_account" "this" {
  name                     = lower("SA0${var.app_name}0${random_id.this.hex}")
  resource_group_name      = azurerm_resource_group.this.name
  location                 = azurerm_resource_group.this.location
  account_tier             = "Standard"
  account_replication_type = "LRS"

  tags = local.tags
}

resource "azurerm_storage_container" "this" {
  name                  = lower("BLB-${var.app_name}")
  storage_account_name  = azurerm_storage_account.this.name
  container_access_type = "private"
}

# For some reason the immutable policy must be treated as an existing resource
resource "azapi_update_resource" "this" {
  type      = "Microsoft.Storage/storageAccounts/blobServices/containers/immutabilityPolicies@2021-08-01"
  name      = "default"
  parent_id = azurerm_storage_container.this.resource_manager_id

  body = jsonencode({
    properties = {
      allowProtectedAppendWrites            = false
      # allowProtectedAppendWritesAll         = null
      immutabilityPeriodSinceCreationInDays = 4
    }
  })
  response_export_values = ["etag"]
}

# This output maybe isn't properly decoded and should be improved
output "etag" {
  value = jsondecode(azapi_update_resource.this.output).etag
}
```

**Comment by ms-henglu** [comment]
Hi @clowa ,

Thanks for the example! About the `etag`, in its GET response, it has the quotes. I have created an issue for that: https://github.com/Azure/azure-rest-api-specs/issues/19134

**Comment by kimsey0** [comment]
We would very much like to be able to create storage accounts with version-level immutability support enabled, since this [can only be set on account creation](https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-policy-configure-version-scope?tabs=azure-cli#enable-support-for-version-level-immutability), not enabled afterwards.

When creating an account with the Azure CLI, version-level immutability is enabled with the `--enable-alw` parameter for the `az storage account create` command.

**Comment by favoretti** [comment]
So I've generated latest stable SDK that would potentially allow us to implement immutability policies (I need those myself ;)). I'll discuss with @tombuildsstuff to agree on an approach and will try take a stab at it.

**Comment by favoretti** [comment]
@kimsey0 As discussed with @tombuildsstuff offline - I'll first add Storage Account-level immutability policy flag and then we'll look into introducing blob-level ones, since that requires a bit more shenanigans.

**Comment by strictness** [comment]
> @tombuildsstuff : do you think if the following kind of REST api around Azure Storage could take part of this [Storage SDK](https://github.com/tombuildsstuff/giovanni)??
> 
> The idea is to be able to create a stored access policy for a given container and then generate a sas key based on this access policy. If it could be managed over Terraform it could facilitate implementations. The main advantage using stored access policies is that we can revoke all generated SAS keys based on a given stored access policy.

That's exactly what we are trying to accomplish. Is there any progress on this implementation?

**Comment by markharveynz** [comment]
We want to implement "VERSION-LEVEL IMMUTABILITY SUPPORT" on a container without a policy. This is required to allow Veeam to use the container and make the storage immutable itself and it cannot have an access policy, rather it needs the tick box done when the container is created. 

Document outlining this is: https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-policy-configure-version-scope?tabs=azure-portal#enable-version-level-immutability-for-a-new-container

**Comment by favoretti** [comment]
@markharveynz by "we want to implement" - do you mean you will be submitting a PR or you are requesting a new enhancement? 

If it's the latter - I'd suggest making a new issue. Will be easier to track progress on it.

**Comment by markharveynz** [comment]
> 

We want to implement the option of "VERSION-LEVEL IMMUTABILITY SUPPORT" on a container as you would when you create a container in the portal, by using Terraform. We do not want to have a policy assigned to the container which is what will happen if you turn on "VERSION-LEVEL IMMUTABILITY SUPPORT" after you create a container as you have to apply a policy.

**Comment by SteveBurkettNZ** [comment]
Interest for this feature is probably being generated by the release of [Veeam Backup & Replication v12](https://www.veeam.com/kb4416) back in February which now supports immutable backups on Azure.

**Comment by DamaniN** [comment]
@markharveynz & @SteveBurkettNZ, I've opened #21512 to have "VERSION-LEVEL IMMUTABILITY SUPPORT"  added to `azurerm_storage_container`. Please vote on it.

**Comment by KaishM** [comment]
Has retention policies (immutability and legal holds) on storage containers been supported now by azurerm_storage_container ? I do not see any PR.

**Comment by philmph** [comment]
For people stumbling across this who want to deploy a container with `Version-level immutability support` enabled **without** assigning a policy (-> setting set at container creation instead of updating it after creation):

On your `azurerm_storage_account` resource add the following block

```hcl
resource "azurerm_storage_account" "this" {

  # ...

  # ID of the resulting "Microsoft.Storage/storageAccounts/blobServices" object is
  # "${azurerm_storage_account.this[0].id}/blobServices/default". "default" is the default name.
  blob_properties {
    versioning_enabled = true
  }
}
```

For the container(s)

```hcl
variable "sa_containers" {
  default     = ["default"]
  description = "List of containers to be created in the Azure Storage Account."
  type        = list(string)
}

resource "azapi_resource" "containers" {
  for_each = toset(var.sa_containers)

  type = "Microsoft.Storage/storageAccounts/blobServices/containers@2022-09-01"
  name      = each.key

  # We append '/blobServices/default' to the storage_account.id see desc. above
  parent_id = "${azurerm_storage_account.this.id}/blobServices/default"

  body = jsonencode({
    properties = {
      immutableStorageWithVersioning = {
        enabled = true
      }
    }
  })
}
```

References used:

[blobServices](https://learn.microsoft.com/en-us/azure/templates/microsoft.storage/storageaccounts/blobservices?pivots=deployment-language-terraform)
[containers](https://learn.microsoft.com/en-us/azure/templates/microsoft.storage/storageaccounts/blobservices/containers?pivots=deployment-language-terraform)
